## Variable Types

Python doesn't generally require *explicit* variable type declarations (with some exceptions that will come later as we get into more advanced programming).  However, it is still useful to know what kinds of data there is, what can be done with it, and how it's stored.

First, let's explore data types like `int`, `float`, `list`, `tuple`, and `string`.


In [1]:
my_int    = 2
my_float  = 3.1415
my_list   = [1,3.1415,"Hello World!","pizza"]
my_tuple  = (5,6,7,8,9)
my_string = "Hello World!"

These examples are fairly simple.  

- `my_int` is an integer, and gets treated like one.  Integers are useful for things like indexes, counters, and so forth.
- `my_float` is a float (often called a "double" in other programming languages), and are regular numbers including decimals.
- `my_list` is a list of values enclosed in square brackets.  Lists are indexed from zero, which means the first item in a list is "item 0".  Lists are great ways to keep collections of data organized and in order, and you can extract individual values simply by including the index with the variable name:  `my_list[3]` will return "pizza".  You can also get values from the end of a list with negative indices.  my_list`[-1]` will return "pizza" because it's the last value.
- `my_tuple` is similar to a list, except that it is a little more difficult to pull individual values from it.  Tuples are useful when you need to maintain groups of values together in relation to each other, such as with (x,y,z) coordinates.
- `my_string` is a list of characters including letters, numbers, punctuation, whitespace (tabs, spaces, line breaks, etc.).  The contents of a string do not include the quotation marks on either side.  Strings can include quotes using *escapes* like `\"` or `\\` to include a backslash.

Variable manipulation comes in many forms and depends on the type of data contained within.  Better understanding of how data types work can allow you to do some interesting things, like taking a "slice" of a string like you would from a list.  

Consider the examples below.

In [2]:
my_list = [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]
# my_list has a length of 26 individual values
len(my_list)

26

In [3]:
my_string = "Once more into the breach!"
# my_string has a length of 26 characters including whitespace and punctuation.
len(my_string)

26

### Slicing Lists

One common use for lists is "slicing", where you can get a small subsection of the list.  Let's say you wanted just the first five elements in `my_list`.  You would use a slice.  Slices are generated similar to how an individual element is called from a list, from inside square brackets.  However, we can put a `:` between the starting and ending indices to get everything between.  We can also use an empty space to indicate "everything".  Check out the examples below.

In [4]:
my_list[:5]

[0, 1, 2, 3, 4]

In [5]:
my_list[5:10]

[5, 6, 7, 8, 9]

Note how the two results are different.  The ending in the first cell is the same as the beginning of the second cell, but we don't actually get "5" in the results in the first cell.  Slices go "up to" the ending value, but don't include it.  Keep this in mind when working with slices.  We can also combine other tricks from list manipulations, like using negative indices to go backwards from the end.

In the next cell, we'll get the last seven elements from the list.

In [6]:
my_list[-7:]

[19, 20, 21, 22, 23, 24, 25]

What if we wanted every third element in the list?

In [7]:
my_list[::3]

[0, 3, 6, 9, 12, 15, 18, 21, 24]

The second `:` indicates a "stride".  This is useful when you have data that is strangely shaped (such as a long list of values that correspond to x,y,z coordinates, but aren't in a (3,n) shaped list.

Now let's combine these.  We'll get every other element starting from the tenth and going up to the twentieth.

In [8]:
my_list[10:20:2]

[10, 12, 14, 16, 18]

Now let's look at strings.  Strings are just lists of letters, numbers, and any other characters you can think of.  With this in mind, we can do things to strings that we have done to lists.

In [9]:
my_string

'Once more into the breach!'

In [10]:
my_string[:5]

'Once '

In [11]:
my_string[5:10]

'more '

In [12]:
my_string[-7:]

'breach!'

In [13]:
my_string[::3]

'Oeo tt eh'

In [14]:
my_string[10:20:2]

'it h '

... some functions are more useful than others, but you get the idea!

Now let's look at integers and floats.  In some programming languages, the difference between these two can be pretty severe.  For example, in C++, dividing a double by an integer will give you a truncated integer, which means you can lose some of the information in your data if you're not careful.  Thankfully, Python is a little more forgiving.

Normal division works like we might intuitively expect, where a float divided by an integer can be a float, and is therefore assumed to be.

In [15]:
my_float/my_int

1.57075

We can also force the division to return an integer value (which is useful in some situations)

In the example above, we got a value of 1.57075.  If we were to round this using conventional methods, we'd get 2
However, forcing integer division with the `//` below gives us a truncated (not rounded) value of 1.0.  This is also slightly deceptive, as the `.0` implies the value is a float,even though the result is a whole number.  This is important to be aware of when doing mathematical work in python.  Truncation just removes everything after the decimal point, while rounding actually considers the value beforehand.


In [16]:
my_float//my_int

1.0

In [17]:
round(my_float/my_int)

2

### Math with Variables

Math can get incredibly complex, so it's important to remember your Order of Operations (PEMDAS) - Parentheses, Exponents, Multiplication, Division, Addition, and Subtraction.

However, in python it's a little different.  Parentheses are solved first, then exponents, until everything in a given equation is reduced down to a series of terms separated by `+`,`-`,`*`, and `/`.  Then, the values are processed left-to-right.

In [18]:
1 + 2 - 3 * 4 / 5

0.6000000000000001

In [19]:
(1 + 2) - 3 * 4 / 5

0.6000000000000001

In [20]:
(1 + 2 - 3 * 4) / 5

-1.8

In [21]:
(1 + 2 - 3) * 4 / 5

0.0

These are just a few examples of how order of operations affects the results.  With this in mind, you can see why it's very important to keep track of what you're doing in a complex mathematical function.  The next cell has a complex equation in a single line, then the same equation separated into more easily-managed terms.

In [22]:
x=3
y=5
z=7
answer =  (x**(y/z)-x/((y+2)*z)-x)/(y*z)*x

print(answer)

-0.07452211053138932


Not only is that difficult to read, but it's also harder to see where errors might be arising.  So we can rewrite it and create additional variables to hold small chunks

In [23]:
x=3
y=5
z=7

# (x**(y/z)-x/((y+2)*z)-x)/(y*z)*x
p = y/z
# (x**p-x/((y+2)*z)-x)/(y*z)*x
q = x**p
# (q-x/((y+2)*z)-x)/(y*z)*x
r = y+2
# (q-x/(r*z)-x)/(y*z)*x
s = r*z
# (q-x/s-x)/(y*z)*x
t = y*z
# (q-x/s-x)/t*x
u = x/s
# (q-u-x)/t*x
v = q-u-x
# v/t*x
w = v/t
# w*x
answer = w*x

print(answer)

-0.07452211053138932


This may seem overengineered, but breaking down the individual terms is helpful in both programming and math, especially when it reveals certain trends, or even ways to rearrange an equation to reduce the overall number of calculations being performed.  This kind of breakdown can also be useful when you begin building larger, more complicated functions, even up to the point of creating entire programs or modules.

### Booleans

Booleans are simply variables that are either `True` or `False`.  They can also be interpreted as `1` and `0`.  Booleans get used all the time in programming, though we may not be constantly aware of them.  

For example, whenever we compare two numbers, the comparison creates a boolean


In [24]:
3<5

True

In [25]:
3>5

False

In [26]:
3 == 5

False

We can see that the responses for the different comparisons are correct.  $3 < 5$ is true, while $3 > 5$ and $3 == 5$ are both false.  Incidentally, the `==` is intentional.  In Python and C++, `=` *assigns* a value, while `==` *compares* two values.

Booleans get used constantly in things like "if-else statements" or "while loops".

### Dictionaries

Another python data type is the `dictionary` (or `dict` as it's written in python).  The dictionary is a very useful datatype, as it can be used to store many different pieces of information in their own types.

In [27]:
# A dictionary is denoted by { } 
my_dictionary = {}

# At this point, "my_dictionary" is an empty dictionary with no keys or values assigned.
# We can assign a key/value pair like this

my_dictionary["Name"] = "Mark"
my_dictionary["Age"]  = 37
my_dictionary["Job"] = "Postdoc"

# Now we can recall any of the values held in the dictionary by using the [key]. Keep in mind, if a key already exists, the previous value will be overwritten.

# If you have a dictionary with keys that you don't know, you can get them like this:
key_list = [key for key in my_dictionary.keys()]
print(key_list)
# This might look strange, but it's done this way because my_dictionary.keys() is a function call that returns an iterative set of single values, rather than the entire list.

# You can also iterate through all the keys and values together.
for key,value in my_dictionary.items():
    print(key,"=",value)



['Name', 'Age', 'Job']
Name = Mark
Age = 37
Job = Postdoc


In [28]:
# In a more relevant example to the lab (and demonstration of dictionary initialization with keys and values):

variant_prmtops = {"WT":"A3H_WT.prmtop",
                   "K121E":"A3H_K121E.prmtop",
                   "K117E":"A3H_K117E.prmtop",
                   "R124D":"A3H_R124D.prmtop"}

# Now I have a list of filenames stored, and I can recall them anytime with this
print(variant_prmtops["K121E"])

# We can also have dictionaries inside dictionaries, which can be useful for bigger datasets.

full_systems = {
"WT" : {"prmtop":"WT.prmtop","trajectory":"WT_100ns.dcd","num_residues":180,"duration":100},
"K121E" : {"prmtop":"K121E.prmtop","trajectory":"K121E_150ns.dcd","num_residues":180,"duration":150},
"K117E" : {"prmtop":"K117E.prmtop","trajectory":"K117E_200ns.dcd","num_residues":180,"duration":200} 
}

print(full_systems["WT"]) ## This prints the entire dictionary

print(full_systems["WT"]["prmtop"])

# You can also store larger datasets inside dictionaries this way.  For example, let's say you have a dataset for the RMSD of an MD trajectory called "rmsd", and one for correlated motion called "correl"
import numpy as np
rmsd = np.random.rand(100)
correl = np.random.rand(50,50)

WT_analyses = {"RMSD":rmsd,"correl":correl}

A3H_K121E.prmtop
{'prmtop': 'WT.prmtop', 'trajectory': 'WT_100ns.dcd', 'num_residues': 180, 'duration': 100}
WT.prmtop


In [29]:
WT_analyses["correl"]

array([[0.07872392, 0.42709779, 0.73786009, ..., 0.14944642, 0.19055872,
        0.45752917],
       [0.14328318, 0.1483461 , 0.63825779, ..., 0.92527165, 0.9000854 ,
        0.16021734],
       [0.50201692, 0.75404792, 0.46468422, ..., 0.44324784, 0.74848111,
        0.94918908],
       ...,
       [0.64425718, 0.47652132, 0.69848477, ..., 0.19183174, 0.42951161,
        0.73094065],
       [0.2925301 , 0.79944765, 0.88059451, ..., 0.65282398, 0.82365397,
        0.56632012],
       [0.68523568, 0.71003483, 0.5788092 , ..., 0.22053449, 0.22295812,
        0.29099571]])

In [30]:
WT_analyses["RMSD"]

array([0.77547714, 0.86825809, 0.73361393, 0.53601816, 0.82347301,
       0.77017413, 0.96120335, 0.41538613, 0.59034859, 0.04939984,
       0.11903879, 0.20902424, 0.76954975, 0.20189964, 0.31954835,
       0.51617544, 0.71262053, 0.44747436, 0.17991915, 0.95935479,
       0.02334348, 0.37246611, 0.37715923, 0.31250376, 0.83167922,
       0.7179388 , 0.26188751, 0.10515804, 0.62818762, 0.82602609,
       0.98213736, 0.22903547, 0.72848045, 0.45872938, 0.26119027,
       0.05973667, 0.65432271, 0.86798405, 0.66082425, 0.0277142 ,
       0.45905896, 0.74384669, 0.29576668, 0.89319424, 0.14992499,
       0.1549802 , 0.89640709, 0.49811178, 0.07505158, 0.85436875,
       0.21030718, 0.13886558, 0.0993692 , 0.04939456, 0.71672669,
       0.00323682, 0.60787663, 0.43602982, 0.31186697, 0.02260092,
       0.41310232, 0.56875889, 0.27089672, 0.24635137, 0.25593297,
       0.65855327, 0.7809256 , 0.37840985, 0.18978786, 0.50569471,
       0.36361048, 0.32712705, 0.5499439 , 0.71682757, 0.01187